src/main/kotlin/org/rust/ide/inspections/RsDeprecationInspection.kt

```diff
@@ -0,0 +1,99 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.ide.inspections
+
+import com.intellij.codeInspection.ProblemHighlightType
+import com.intellij.codeInspection.ProblemsHolder
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiElementVisitor
+import com.intellij.psi.util.parents
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.*
+
+class RsDeprecationInspection : RsLocalInspectionTool() {
+    override fun getDisplayName() = "Deprecated item"
+
+    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean): PsiElementVisitor = object : RsVisitor() {
+        override fun visitElement(ref: RsElement) {
+            // item is non-inline module declaration or not reference element
+            if (ref is RsModDeclItem || ref !is RsWeakReferenceElement) return
+
+            var original = ref.reference?.resolve() ?: return
+
+            val identifier = ref.referenceNameElement ?: ref
+            original = when (original) {
+                is RsFile -> original.declaration
+                is RsAbstractable -> original.superItem
```

Comment:
We should call `superItem` only when `RsAbstractable.owner.isImplOrTrait`

Reply:
Done, but we need fallback like this: 
```kotlin
... original.owner.isImplOrTrait -> original.superItem ?: original
```

Otherwise, this case doesn't work:
```rust
struct Foo;

impl Foo {
    #[deprecated]
    fn new() -> Foo {
        Foo { }
    }
}

fn main() {
    let foo = Foo::<warning descr="`new` is deprecated">new</warning>();
}
```

Reply:
Answered in https://github.com/intellij-rust/intellij-rust/pull/3550#discussion_r268511363